include/mf_nh_operator.h

```diff
@@ -1210,34 +1180,23 @@ NeoHookOperator<dim, fe_degree, n_q_points_1d, Number>::compute_diagonal()
   VectorType &inverse_diagonal_vector = inverse_diagonal_entries->get_vector();
   VectorType &diagonal_vector         = diagonal_entries->get_vector();
 
-  data_current->initialize_dof_vector(inverse_diagonal_vector);
-  data_current->initialize_dof_vector(diagonal_vector);
-
-  unsigned int dummy = 0;
-  diagonal_vector    = 0.;
-  local_diagonal_cell(*data_current,
-                      diagonal_vector,
-                      dummy,
-                      std::make_pair<unsigned int, unsigned int>(
-                        0, data_current->n_cell_batches()));
-  diagonal_vector.compress(VectorOperation::add);
-
-  // data_current->cell_loop (&NeoHookOperator::local_diagonal_cell,
-  //                          this, diagonal_vector, dummy);
+  MatrixFreeTools::compute_diagonal(*data_in_use,
+                                    diagonal_vector,
+                                    &NeoHookOperator::do_operation_on_cell,
+                                    this);
 
+  data_in_use->initialize_dof_vector(inverse_diagonal_vector);
   // set_constrained_entries_to_one
   Assert(data_current->get_constrained_dofs().size() ==
            data_reference->get_constrained_dofs().size(),
          ExcInternalError());
   for (const auto dof : data_current->get_constrained_dofs())
     diagonal_vector.local_element(dof) = 1.;
```

Comment:
@peterrum this is still required, right?
BTW your tool implementation is more than twice as slow as the implementation here. This was a bit unexpected.

Reply:
> this is still required, right?

Yes. Thos values are undefined (set to zero).

> BTW your tool implementation is more than twice as slow as the implementation here. This was a bit unexpected.

I know. This is the cost for correctness and explicit operations on lanes (like  https://github.com/dealii/dealii/blob/c0b01e2791e99632132e68c6f78fa44c57d79d9e/include/deal.II/matrix_free/tools.h#L541-L578 ) ;)

Reply:
I guess one could optimize the code for standard containment cases...

Reply:
Alright, thanks!

Reply:
Maybe `vectorization_category_strict` would help to bypass the explicit code snippet. I only have two types of constraints here (DRB only one).